### ISSUE TITLE
Content Pipeline array serialization issue

### ISSUE THREAD

**Issue by Eversor** [MonoGame/MonoGame#3399]
"Content reader could not be found for System.Int32 type." when i try to load a custom processed array of items. It seems like the Pipeline associates 2 types of readers, the first one is correct while the second one is a Reflective Reader whose initializator throws complaining it cant find the System.Int32 reader to read a field of my object. I don't know yet if this happens on single content types.
By modifying the pipeline to catch the error i can go past the error and the elements get loaded just fine.

**Comment by tomspilman** [comment]
So this is a bug at runtime in your game?

Can you give me some more specifics about what this "custom processed array of items" is?

Also are you running a development build or the 3.2 release?

**Comment by Eversor** [comment]
It seems more like an error of the pipeline, an array of items is just an array of instances of the same class (say type Enemy[]) that is meant to be loaded from a single file (say an XML in my case)

**Comment by tomspilman** [comment]
I don't follow what the problem is.

I need specifics to be able to fix this.  What specifically is the type you're trying to de-serialize?

**Comment by Eversor** [comment]
Of course, i currently don't have my repositories at hand due to an hard drive failure (installing visual studio as i write :( ).  I'm picking the files and reducing them to usable size here. Once i finsh installing the visual studio ce13 will provide more readable stuff but it might help if you have spare time now.

The following works fine in xna:
--I have an xml : 

<ArrayOfEnemy>
 <Enemy>
  <agi>15</agi>
  <maxhp>30</maxhp>
  <spi>10</spi>
  <gold>2</gold>
 </Enemy>
 <Enemy>
  <agi>13</agi>
  <maxhp>50</maxhp>
  <spi>5</spi>
  <gold>4</gold>
 </Enemy>
</ArrayOfEnemy>

--And process  as follows

namespace DeDProcessorLib
{
    [ContentImporter(".xml", DefaultProcessor = XMLEnemysProcessor", DisplayName = "XML Enemys importer")]
    class XMLEnemyImporter : ContentImporter<Enemy[]>
    {
        public override Enemy[] Import(string filename,
            ContentImporterContext context)
        {
            var xml = new XmlSerializer(typeof(Enemy[]));
            FileStream fs=new FileStream(filename,FileMode.Open,FileAccess.Read);
            Enemy[] m = (Enemy[])xml.Deserialize(fs);
            fs.Close();
            return m;
        }
    }

    [ContentProcessor(DisplayName = "XML Enemys Processor")]
    class XMLEnemysProcessor : ContentProcessor<Enemy[], Enemy[]>
    {
        public override Enemy[] Process(Enemy[] input,
            ContentProcessorContext context)
        	{
		//DO some processing here
		}
	}
    [ContentTypeWriter]
    class EnemysWriter : ContentTypeWriter<Enemy[]>
    {
        protected override void Write(ContentWriter output, Enemy[] value)
        {
            output.Write(value.Length);
            for (int i = 0; i < value.Length; i++)
                value[i].binary_write(output);
        }
        public override string GetRuntimeType(TargetPlatform targetPlatform)
        {
            return typeof(Enemy[]).AssemblyQualifiedName;
        }
        public override string GetRuntimeReader(TargetPlatform targetPlatform)
        {
            return "ProjectOne.CustomEnemies, DoomAndDestiny," +
                " Version=1.0.0.0, Culture=neutral";
        }
    }
}

-- in the deployed assembly
namespace ProjectOne
{
    class CustomEnemies : ContentTypeReader<Enemy[]>
    {
        protected override Enemy[] Read(ContentReader input,
            Enemy[] existingInstance)
            {
            if (existingInstance == null)
            {
                existingInstance = Enemy.LoadArray(input);
            }
            return existingInstance;
           }
    }
}

**Comment by Eversor** [comment]
Here is the project, hope it's not to difficult to undesrstand:
1) Open sln
2) Relink missing projs
3) Build the custom pipeline (it should compile already in the ../ContentProj/lib folder
4) Build content project (../ContentProj/lib with the Pipeline Tool )
5) Run the game, voilà
(if you want to skip 1-4 i already put resulting xnb in contentfolder)

https://www.dropbox.com/s/fdlgke3k292qv6l/CustomPipelineArray_Issue.zip?dl=0

**Comment by Eversor** [comment]
I'm doing some inspection on the issue inside the MGCB, it seems like the reflective writer is added as well, my best guess to solve the problem is to catch the ReflectiveReader initialization Exception and retrow only if the ReflectiveReader is the actual selected reader. What do you think?

**Comment by tomspilman** [comment]
> my best guess to solve the problem is to catch the ReflectiveReader initialization Exception and retrow only if the ReflectiveReader is the actual selected reader

That doesn't sound like the right fix to me.  I or someone else will have to find time to convert you're example project into a unit test and then develop a fix.

**Comment by tgjones** [comment]
I'll have a look at this.

**Comment by tgjones** [comment]
So based on @Eversor's test project, I've written a failing test for this:
https://github.com/tgjones/MonoGame/blob/contentpipeline-item-array-bug/Test/ContentPipeline/ContentTypeWriterReaderTests.cs

The problem is fairly complex. It manifests as a crash when loading content, but the root cause is actually much earlier, when writing content.

We call `ContentWriter.GetTypeWriter` for `TestEntity[]`. Then, because of [this](https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework.Content.Pipeline/Serialization/Compiler/ContentWriter.cs#L254):

``` csharp
else if (type.IsArray)
{
    GetTypeWriter(type.GetElementType());
}
```

... we recursively call `GetTypeWriter` for `TestEntity`. This creates a `ReflectiveWriter<TestEntity>`.

To test whether this is correct, I've saved the `.xnb` generated by XNA for the `TestEntity[]` test, and loaded it up using MG's content pipeline. And... XNA only writes out one content type reader, namely the custom `TestEntityArrayReader` that I would expect.

But MG writes out two content type readers to the `.xnb` file: `TestEntityArrayReader`, and also `ReflectiveReader<TestEntity>` (because of the recursive call to `GetTypeWriter` when `type.IsArray`).

I guess the "TODO: This is kinda messy" comment immediately above those lines in `ContentWriter` was a foreshadowing of this ;-) @tomspilman, do you have any suggestions for how to proceed? If not, no worries, I'll dig deeper.

As a first attempt, I simply commented out the recursive `GetTypeWriter` call for the array element type, which actually made this test pass, but several other tests in `IntermediateDeserializerTest` failed, because the element type reader isn't written out to the `.xnb`.

To handle both cases - (1) `GetTypeWriter(typeof(SomeElement[]))` where we will not need a content type writer for `SomeElement` (which is what this issue is about) and (2) `GetTypeWriter(typeof(SomeElement[]))` where we WILL need a content type writer for `SomeElement` - is a little tricky. `ArrayWriter.Write` calls `ContentTypeWriter.WriteObject(element, _elementWriter)`, and because we pass `_elementWriter`, the logic in the other overload of `WriteObject`, which would register the necessary content type writer, is bypassed. (Why is this? Why do we need to pass `_elementWriter` - wouldn't `WriteObject(element)` be able to compute it?)

**Comment by tomspilman** [comment]
> But MG writes out two content type readers to the  .xnb  file:  TestEntityArrayReader , and also  ReflectiveReader<TestEntity> 

How does `TestEntityArrayReader` get added?  Is it added first before the `ReflectiveReader<TestEntity>`?


> I simply commented out the recursive  GetTypeWriter  call for the array element type, which actually made this test pass, but several other tests in  IntermediateDeserializerTest  failed

Yeah...  that won't work.

It seems like what we need to do here is to detect if we already have a customized reader (`TestEntityArrayReader` in this case) that can manage the `TestEntity[]` type.

So the code there would be more like...

```
else if (type.IsArray && !HasExistingWriterFor(type))
{
    GetTypeWriter(type.GetElementType());
}
```

This assumes that `TestEntityArrayReader` gets added first.


> Why do we need to pass  _elementWriter  - wouldn't  WriteObject(element)  be able to compute it?

That code is pretty complex...  i'm not sure.  It kinda evolved to where it is from adding tests and getting them to pass.

**Comment by tgjones** [comment]
> How does TestEntityArrayReader get added? Is it added first before the ReflectiveReader<TestEntity>?

Yes - it's something like this:

1. `ContentWriter.GetTypeWriter(typeof(TestEntity[]))` is called.
2. That first calls `ContentCompiler.GetTypeWriter(typeof(TestEntity[]))`, which returns `TestEntityArrayReader`, which gets added to the `typeMap`.
3. Then, because of `if (type.IsArray)`, it calls `GetTypeWriter(typeof(TestEntity))`.
4. That calls `ContentCompiler.GetTypeWriter(typeof(TestEntity))`, which returns `ReflectiveReader<TestEntity>`, which gets added to the `typeMap`.

So the problem isn't with detecting that we have a customised reader for `TestEntity[]` - that works fine - it's that we shouldn't *always* add a type writer for `TestEntity` (the array element type). (Sometimes we should - but I think, only when we specifically want the type writer for `TestEntity`, not as a side-effect of getting the type writer for `TestEntity[]`. If `ArrayWriter.Write` called `ContentTypeWriter.WriteObject(element)`, without passing `_elementWriter`, it seems like that would help, because it would result in a call to `ContentWriter.GetTypeWriter(typeof(TestEntity))` at the point where it's needed.

I'll try to find some time to have a deeper look at this, and see if simply changing `ArrayWriter.Write` helps. I suspect it's more complicated than that though.

**Comment by SimonDarksideJ** [comment]
I cannot replicate this issue on the latest 3.8.4.1 release, so I am closing this issue. Please reopen if it is still and issue
